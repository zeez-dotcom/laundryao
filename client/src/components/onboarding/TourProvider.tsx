import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState,
  type ReactNode,
} from "react";
import { Dialog, DialogContent, DialogDescription, DialogFooter, DialogHeader, DialogTitle } from "@/components/ui/dialog";
import { Button } from "@/components/ui/button";
import { Progress } from "@/components/ui/progress";
import { Tooltip, TooltipContent, TooltipTrigger } from "@/components/ui/tooltip";
import { cn } from "@/lib/utils";

const STORAGE_KEYS = {
  dismissed: "laundryao.tours.dismissed",
  progress: "laundryao.tours.progress",
  glossary: "laundryao.tours.glossary",
};

type TourStep = {
  id: string;
  title: string;
  description: string;
  ctaLabel?: string;
};

type TourDefinition = {
  id: string;
  title: string;
  description?: string;
  steps: TourStep[];
};

type GlossaryEntry = {
  term: string;
  description: string;
};

type TourContextValue = {
  registerTour: (tour: TourDefinition) => () => void;
  startTour: (tourId: string) => void;
  goToStep: (step: number) => void;
  completeStep: () => void;
  dismissTour: (tourId: string) => void;
  isTourDismissed: (tourId: string) => boolean;
  activeTourId: string | null;
  activeStepIndex: number;
  activeTour?: TourDefinition;
  registerGlossaryEntries: (entries: GlossaryEntry[]) => void;
  glossary: Record<string, GlossaryEntry>;
  seenGlossaryTerms: Set<string>;
  markGlossarySeen: (term: string) => void;
  availableTours: TourDefinition[];
};

const TourContext = createContext<TourContextValue | null>(null);

function readLocalStorage<T>(key: string, fallback: T): T {
  if (typeof window === "undefined") return fallback;
  try {
    const value = window.localStorage.getItem(key);
    if (!value) return fallback;
    return JSON.parse(value) as T;
  } catch (error) {
    console.warn(`Failed to read local storage for ${key}`, error);
    return fallback;
  }
}

function writeLocalStorage<T>(key: string, value: T) {
  if (typeof window === "undefined") return;
  try {
    window.localStorage.setItem(key, JSON.stringify(value));
  } catch (error) {
    console.warn(`Failed to persist ${key}`, error);
  }
}

export function TourProvider({ children }: { children: ReactNode }) {
  const [tours, setTours] = useState<Record<string, TourDefinition>>({});
  const [activeTourId, setActiveTourId] = useState<string | null>(null);
  const [activeStepIndex, setActiveStepIndex] = useState(0);
  const [dismissedTours, setDismissedTours] = useState<string[]>(() => readLocalStorage(STORAGE_KEYS.dismissed, []));
  const [tourProgress, setTourProgress] = useState<Record<string, number>>(() => readLocalStorage(STORAGE_KEYS.progress, {}));
  const [glossary, setGlossary] = useState<Record<string, GlossaryEntry>>(() => readLocalStorage(STORAGE_KEYS.glossary, {}));
  const [seenGlossaryTerms, setSeenGlossaryTerms] = useState<Set<string>>(() => {
    const initial = readLocalStorage<string[]>(`${STORAGE_KEYS.glossary}.seen`, []);
    return new Set(initial);
  });
  const isDismissingRef = useRef(false);

  useEffect(() => {
    writeLocalStorage(STORAGE_KEYS.dismissed, dismissedTours);
  }, [dismissedTours]);

  useEffect(() => {
    writeLocalStorage(STORAGE_KEYS.progress, tourProgress);
  }, [tourProgress]);

  useEffect(() => {
    writeLocalStorage(STORAGE_KEYS.glossary, glossary);
  }, [glossary]);

  useEffect(() => {
    writeLocalStorage(`${STORAGE_KEYS.glossary}.seen`, Array.from(seenGlossaryTerms));
  }, [seenGlossaryTerms]);

  const registerTour = useCallback((tour: TourDefinition) => {
    setTours((current) => ({ ...current, [tour.id]: tour }));
    return () => {
      setTours((current) => {
        const next = { ...current };
        delete next[tour.id];
        return next;
      });
    };
  }, []);

  const startTour = useCallback(
    (tourId: string) => {
      if (dismissedTours.includes(tourId)) return;
      const tour = tours[tourId];
      if (!tour || tour.steps.length === 0) return;
      setActiveTourId(tourId);
      setActiveStepIndex(() => {
        const savedStep = tourProgress[tourId] ?? 0;
        return Math.min(savedStep, tour.steps.length - 1);
      });
    },
    [dismissedTours, tours, tourProgress],
  );

  const goToStep = useCallback(
    (index: number) => {
      if (!activeTourId) return;
      const tour = tours[activeTourId];
      if (!tour) return;
      const clamped = Math.max(0, Math.min(index, tour.steps.length - 1));
      setActiveStepIndex(clamped);
      setTourProgress((current) => ({ ...current, [activeTourId]: clamped }));
    },
    [activeTourId, tours],
  );

  const dismissTour = useCallback(
    (tourId: string) => {
      setDismissedTours((current) => (current.includes(tourId) ? current : [...current, tourId]));
      setTourProgress((current) => ({ ...current, [tourId]: 0 }));
      if (activeTourId === tourId) {
        isDismissingRef.current = true;
        setActiveTourId(null);
      }
    },
    [activeTourId],
  );

  const completeStep = useCallback(() => {
    if (!activeTourId) return;
    const tour = tours[activeTourId];
    if (!tour) return;
    const nextIndex = activeStepIndex + 1;
    if (nextIndex >= tour.steps.length) {
      dismissTour(tour.id);
      return;
    }
    setActiveStepIndex(nextIndex);
    setTourProgress((current) => ({ ...current, [tour.id]: nextIndex }));
  }, [activeTourId, activeStepIndex, dismissTour, tours]);

  useEffect(() => {
    if (!activeTourId || isDismissingRef.current) {
      isDismissingRef.current = false;
      return;
    }
    const tour = tours[activeTourId];
    if (!tour) {
      setActiveTourId(null);
      return;
    }
    setTourProgress((current) => ({ ...current, [tour.id]: activeStepIndex }));
  }, [activeTourId, activeStepIndex, tours]);

  const registerGlossaryEntries = useCallback((entries: GlossaryEntry[]) => {
    setGlossary((current) => {
      const next = { ...current };
      entries.forEach((entry) => {
        const key = entry.term.toLowerCase();
        next[key] = entry;
      });
      return next;
    });
  }, []);

  const markGlossarySeen = useCallback((term: string) => {
    setSeenGlossaryTerms((current) => {
      const next = new Set(current);
      next.add(term.toLowerCase());
      return next;
    });
  }, []);

  const isTourDismissed = useCallback((tourId: string) => dismissedTours.includes(tourId), [dismissedTours]);

  const activeTour = activeTourId ? tours[activeTourId] : undefined;

  const value = useMemo<TourContextValue>(
    () => ({
      registerTour,
      startTour,
      goToStep,
      completeStep,
      dismissTour,
      isTourDismissed,
      activeTourId,
      activeStepIndex,
      activeTour,
      registerGlossaryEntries,
      glossary,
      seenGlossaryTerms,
      markGlossarySeen,
      availableTours: Object.values(tours),
    }),
    [
      registerTour,
      startTour,
      goToStep,
      completeStep,
      dismissTour,
      isTourDismissed,
      activeTourId,
      activeStepIndex,
      activeTour,
      registerGlossaryEntries,
      glossary,
      seenGlossaryTerms,
      markGlossarySeen,
      tours,
    ],
  );

  const currentStep = activeTour ? activeTour.steps[activeStepIndex] : undefined;
  const progress = activeTour ? ((activeStepIndex + 1) / activeTour.steps.length) * 100 : 0;

  return (
    <TourContext.Provider value={value}>
      {children}
      <Dialog open={Boolean(activeTour)} onOpenChange={(open) => !open && activeTour && dismissTour(activeTour.id)}>
        <DialogContent aria-describedby={currentStep ? `${currentStep.id}-description` : undefined}>
          {activeTour ? (
            <>
              <DialogHeader>
                <DialogTitle className="font-heading text-[var(--text-lg)]">{activeTour.title}</DialogTitle>
                <DialogDescription className="text-[var(--text-sm)]">
                  {activeTour.description || "Follow each step to get oriented."}
                </DialogDescription>
              </DialogHeader>
              {currentStep ? (
                <div className="space-y-4">
                  <div className="rounded-lg bg-[var(--surface-muted)] p-4">
                    <div className="text-[var(--text-sm)] font-semibold" id={`${currentStep.id}-title`}>
                      {currentStep.title}
                    </div>
                    <p id={`${currentStep.id}-description`} className="text-[var(--text-sm)] text-muted-foreground">
                      {currentStep.description}
                    </p>
                  </div>
                  <Progress value={progress} aria-hidden={false} aria-label="Onboarding progress" />
                </div>
              ) : null}
              <DialogFooter className="gap-2">
                <Button variant="ghost" onClick={() => activeTour && dismissTour(activeTour.id)}>
                  Skip tour
                </Button>
                {activeTour && currentStep ? (
                  <Button onClick={completeStep}>{currentStep.ctaLabel || "Next"}</Button>
                ) : null}
              </DialogFooter>
            </>
          ) : null}
        </DialogContent>
      </Dialog>
    </TourContext.Provider>
  );
}

export function useTour() {
  const context = useContext(TourContext);
  if (!context) {
    throw new Error("useTour must be used within a TourProvider");
  }
  return context;
}

type GlossaryTooltipProps = {
  term: string;
  children?: ReactNode;
  className?: string;
};

export function GlossaryTooltip({ term, children, className }: GlossaryTooltipProps) {
  const { glossary, markGlossarySeen, seenGlossaryTerms } = useTour();
  const key = term.toLowerCase();
  const entry = glossary[key];

  if (!entry) {
    return <>{children ?? term}</>;
  }

  const hasSeen = seenGlossaryTerms.has(key);

  return (
    <Tooltip>
      <TooltipTrigger asChild>
        <button
          type="button"
          onFocus={() => markGlossarySeen(term)}
          onMouseEnter={() => markGlossarySeen(term)}
          className={cn(
            "inline-flex items-center gap-1 rounded-full px-2 py-1 text-[var(--text-xs)]",
            hasSeen ? "bg-[var(--surface-muted)] text-muted-foreground" : "bg-primary/10 text-primary",
            className,
          )}
        >
          {children ?? entry.term}
        </button>
      </TooltipTrigger>
      <TooltipContent className="max-w-xs space-y-2">
        <div className="text-[var(--text-sm)] font-semibold">{entry.term}</div>
        <p className="text-[var(--text-xs)] text-muted-foreground">{entry.description}</p>
        <Button variant="ghost" size="sm" onClick={() => markGlossarySeen(term)}>
          Got it
        </Button>
      </TooltipContent>
    </Tooltip>
  );
}
