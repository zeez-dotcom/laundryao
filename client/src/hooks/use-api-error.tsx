import { useMemo } from "react";
import { AlertCircle } from "lucide-react";
import EmptyState from "@/components/common/EmptyState";
import { useTranslation } from "@/lib/i18n";

export function useApiError(error: unknown) {
  const { t } = useTranslation();
  return useMemo(() => {
    if (!error) return null;
    const message = error instanceof Error ? error.message : String(error);
    return (
      <EmptyState
        icon={<AlertCircle className="w-12 h-12 text-red-500 mb-4" />}
        title={t.error}
        description={message}
      />
    );
  }, [error, t]);
}
