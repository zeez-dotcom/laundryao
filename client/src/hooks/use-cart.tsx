import { useState, useCallback } from "react";
import type { CartItem, CartSummary } from "@shared/schema";
import { getTaxRate } from "@/lib/tax";

export function useCart() {
  const [cartItems, setCartItems] = useState<CartItem[]>([]);
  const [paymentMethod, setPaymentMethod] = useState<"cash" | "card">("cash");

  const addToCart = useCallback((product: { id: string; name: string; price: string; imageUrl?: string }) => {
    setCartItems(prev => {
      const existing = prev.find(item => item.id === product.id);
      if (existing) {
        return prev.map(item =>
          item.id === product.id
            ? { ...item, quantity: item.quantity + 1, total: (item.quantity + 1) * item.price }
            : item
        );
      }

      const price = parseFloat(product.price);
      return [...prev, {
        id: product.id,
        name: product.name,
        price,
        quantity: 1,
        total: price,
        imageUrl: product.imageUrl
      }];
    });
  }, []);

  const updateQuantity = useCallback((id: string, quantity: number) => {
    if (quantity <= 0) {
      setCartItems(prev => prev.filter(item => item.id !== id));
      return;
    }
    
    setCartItems(prev =>
      prev.map(item =>
        item.id === id
          ? { ...item, quantity, total: quantity * item.price }
          : item
      )
    );
  }, []);

  const removeFromCart = useCallback((id: string) => {
    setCartItems(prev => prev.filter(item => item.id !== id));
  }, []);

  const clearCart = useCallback(() => {
    setCartItems([]);
  }, []);

  const getCartSummary = useCallback((): CartSummary => {
    const subtotal = cartItems.reduce((sum, item) => sum + item.total, 0);
    const taxRate = getTaxRate();
    const tax = subtotal * taxRate;
    const total = subtotal + tax;
    const itemCount = cartItems.reduce((sum, item) => sum + item.quantity, 0);

    return {
      items: cartItems,
      subtotal: Math.round(subtotal * 100) / 100,
      tax: Math.round(tax * 100) / 100,
      total: Math.round(total * 100) / 100,
      itemCount
    };
  }, [cartItems]);

  return {
    cartItems,
    paymentMethod,
    setPaymentMethod,
    addToCart,
    updateQuantity,
    removeFromCart,
    clearCart,
    getCartSummary
  };
}
