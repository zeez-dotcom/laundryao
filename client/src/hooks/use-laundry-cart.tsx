import { useState, useCallback } from "react";
import type {
  LaundryCartItem,
  LaundryCartSummary,
  ClothingItem,
  LaundryServiceWithItemPrice,
} from "@shared/schema";
import { getTaxRate } from "@/lib/tax";
import { apiRequest } from "@/lib/queryClient";

type AppliedCoupon = {
  id: string;
  code: string;
  discountType: "percentage" | "fixed";
  discountValue: number;
  discount: number;
  nameEn: string;
  applicationType: "whole_cart" | "specific_items" | "specific_services";
  applicableItems?: LaundryCartItem[];
};

export function useLaundryCart() {
  const [cartItems, setCartItems] = useState<LaundryCartItem[]>([]);
  const [paymentMethod, setPaymentMethod] = useState<"cash" | "card" | "pay_later">("cash");
  const [appliedCoupon, setAppliedCoupon] = useState<AppliedCoupon | null>(null);
  const [couponCode, setCouponCode] = useState("");
  const [isCouponLoading, setIsCouponLoading] = useState(false);
  const [couponError, setCouponError] = useState<string | null>(null);

  const addToCart = useCallback((clothingItem: ClothingItem, service: LaundryServiceWithItemPrice, quantity: number = 1) => {
    setCartItems(prev => {
      // Create unique ID combining clothing item and service
      const uniqueId = `${clothingItem.id}-${service.id}`;
      const existing = prev.find(item => item.id === uniqueId);

      if (existing) {
        return prev.map(item =>
          item.id === uniqueId
            ? {
                ...item,
                quantity: item.quantity + quantity,
                total: (item.quantity + quantity) * parseFloat(service.itemPrice ?? service.price)
              }
            : item
        );
      }

      const price = parseFloat(service.itemPrice ?? service.price);
      return [...prev, {
        id: uniqueId,
        clothingItem,
        service,
        quantity,
        total: quantity * price
      }];
    });
  }, []);

  const updateQuantity = useCallback((id: string, quantity: number) => {
    if (quantity <= 0) {
      setCartItems(prev => prev.filter(item => item.id !== id));
      return;
    }
    
    setCartItems(prev =>
      prev.map(item =>
        item.id === id
          ? {
              ...item,
              quantity,
              total: quantity * parseFloat(item.service.itemPrice ?? item.service.price),
            }
          : item
      )
    );
  }, []);

  const removeFromCart = useCallback((id: string) => {
    setCartItems(prev => prev.filter(item => item.id !== id));
  }, []);

  const clearCart = useCallback(() => {
    setCartItems([]);
    setAppliedCoupon(null);
    setCouponCode("");
    setCouponError(null);
  }, []);

  const applyCoupon = useCallback(async (code: string, branchId: string) => {
    setIsCouponLoading(true);
    setCouponError(null);
    
    try {
      const cartData = cartItems.map(item => ({
        clothingItem: item.clothingItem,
        service: item.service,
        quantity: item.quantity,
        total: item.total
      }));

      const response = await apiRequest("POST", "/api/coupons/validate", {
        code: code.trim(),
        branchId,
        cartItems: cartData
      });

      const result = await response.json();

      if (!response.ok) {
        throw new Error(result.message || "Failed to validate coupon");
      }

      if (result.valid && result.coupon) {
        const coupon: AppliedCoupon = {
          id: result.coupon.id,
          code: result.coupon.code,
          discountType: result.coupon.discountType,
          discountValue: parseFloat(result.coupon.discountValue),
          discount: result.discount,
          nameEn: result.coupon.nameEn,
          applicationType: result.coupon.applicationType,
          applicableItems: result.applicableItems || cartItems
        };

        setAppliedCoupon(coupon);
        setCouponCode(code.trim());
        return { success: true, coupon };
      } else {
        throw new Error(result.message || "Invalid coupon");
      }
    } catch (error: any) {
      setCouponError(error.message);
      return { success: false, error: error.message };
    } finally {
      setIsCouponLoading(false);
    }
  }, [cartItems]);

  const removeCoupon = useCallback(() => {
    setAppliedCoupon(null);
    setCouponCode("");
    setCouponError(null);
  }, []);

  const getCartSummary = useCallback((): LaundryCartSummary & { 
    coupon?: AppliedCoupon; 
    discountAmount?: number; 
    totalBeforeDiscount?: number; 
  } => {
    const subtotal = cartItems.reduce((sum, item) => sum + item.total, 0);
    const taxRate = getTaxRate();
    
    let discountAmount = 0;
    let applicableSubtotal = subtotal;

    // Apply coupon discount if available
    if (appliedCoupon) {
      if (appliedCoupon.applicationType === "whole_cart") {
        applicableSubtotal = subtotal;
      } else if (appliedCoupon.applicableItems && appliedCoupon.applicableItems.length > 0) {
        applicableSubtotal = appliedCoupon.applicableItems.reduce((sum, item) => sum + item.total, 0);
      }
      discountAmount = Math.min(appliedCoupon.discount, applicableSubtotal);
    }

    const discountedSubtotal = subtotal - discountAmount;
    const tax = discountedSubtotal * taxRate;
    const total = discountedSubtotal + tax;
    const itemCount = cartItems.reduce((sum, item) => sum + item.quantity, 0);

    return {
      items: cartItems,
      subtotal: Math.round(subtotal * 100) / 100,
      tax: Math.round(tax * 100) / 100,
      total: Math.round(total * 100) / 100,
      itemCount,
      coupon: appliedCoupon || undefined,
      discountAmount: Math.round(discountAmount * 100) / 100,
      totalBeforeDiscount: appliedCoupon ? Math.round((subtotal + (subtotal * taxRate)) * 100) / 100 : undefined
    };
  }, [cartItems, appliedCoupon]);

  return {
    cartItems,
    paymentMethod,
    setPaymentMethod,
    addToCart,
    updateQuantity,
    removeFromCart,
    clearCart,
    getCartSummary,
    // Coupon functionality
    appliedCoupon,
    couponCode,
    isCouponLoading,
    couponError,
    applyCoupon,
    removeCoupon,
    setCouponCode
  };
}
