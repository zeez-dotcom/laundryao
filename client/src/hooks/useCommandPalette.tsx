import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
  type DependencyList,
  type ReactNode,
} from "react";

type CommandAction = {
  id: string;
  label: string;
  description?: string;
  group?: string;
  keywords?: string[];
  shortcut?: string;
  run: () => void;
};

type CommandPaletteContextValue = {
  isOpen: boolean;
  setOpen: (open: boolean) => void;
  toggle: () => void;
  commands: CommandAction[];
  registerCommands: (commands: CommandAction[]) => () => void;
  runCommand: (id: string) => void;
};

const CommandPaletteContext = createContext<CommandPaletteContextValue | null>(null);

function normalizeCommands(commands: CommandAction[]) {
  const map = new Map<string, CommandAction>();
  commands.forEach((command) => {
    map.set(command.id, command);
  });
  return Array.from(map.values());
}

export function CommandPaletteProvider({ children }: { children: ReactNode }) {
  const [isOpen, setIsOpen] = useState(false);
  const [commands, setCommands] = useState<CommandAction[]>([]);

  const registerCommands = useCallback((newCommands: CommandAction[]) => {
    setCommands((existing) => {
      return normalizeCommands([...existing, ...newCommands]);
    });

    return () => {
      setCommands((existing) => existing.filter((command) => !newCommands.some((item) => item.id === command.id)));
    };
  }, []);

  const setOpen = useCallback((open: boolean) => {
    setIsOpen(open);
  }, []);

  const toggle = useCallback(() => {
    setIsOpen((prev) => !prev);
  }, []);

  const runCommand = useCallback(
    (id: string) => {
      const command = commands.find((item) => item.id === id);
      if (!command) return;
      command.run();
      setIsOpen(false);
    },
    [commands],
  );

  useEffect(() => {
    const handleKey = (event: KeyboardEvent) => {
      const key = event.key.toLowerCase();
      if ((event.metaKey || event.ctrlKey) && !event.shiftKey && key === "k") {
        event.preventDefault();
        toggle();
      } else if ((event.metaKey || event.ctrlKey) && event.shiftKey && (key === "p" || key === " ")) {
        event.preventDefault();
        toggle();
      } else if (event.key === "Escape") {
        setIsOpen(false);
      }
    };
    window.addEventListener("keydown", handleKey);
    return () => {
      window.removeEventListener("keydown", handleKey);
    };
  }, [toggle]);

  const value = useMemo<CommandPaletteContextValue>(
    () => ({
      isOpen,
      setOpen,
      toggle,
      commands,
      registerCommands,
      runCommand,
    }),
    [commands, isOpen, registerCommands, runCommand, toggle],
  );

  return <CommandPaletteContext.Provider value={value}>{children}</CommandPaletteContext.Provider>;
}

export function useCommandPalette() {
  const context = useContext(CommandPaletteContext);
  if (!context) {
    throw new Error("useCommandPalette must be used within CommandPaletteProvider");
  }
  return context;
}

export function useCommandRegistration(commands: CommandAction[], deps: DependencyList = []) {
  const { registerCommands } = useCommandPalette();
  useEffect(() => {
    const unregister = registerCommands(commands);
    return () => unregister();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [registerCommands, ...deps]);
}

export type { CommandAction };
