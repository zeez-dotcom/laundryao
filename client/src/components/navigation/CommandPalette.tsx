import { useEffect, useMemo } from "react";
import {
  CommandDialog,
  CommandEmpty,
  CommandGroup,
  CommandInput,
  CommandItem,
  CommandList,
  CommandSeparator,
  CommandShortcut,
} from "@/components/ui/command";
import { useCommandPalette } from "@/hooks/useCommandPalette";
import { useLocation } from "wouter";
import { useTour } from "@/components/onboarding/TourProvider";
import { LifeBuoy, ListChecks, Package, Settings2, ShoppingCart, Users } from "lucide-react";

type GroupedCommand = {
  heading: string;
  items: {
    id: string;
    label: string;
    description?: string;
    shortcut?: string;
  }[];
};

export function CommandPalette() {
  const { isOpen, setOpen, registerCommands, commands, runCommand } = useCommandPalette();
  const [, setLocation] = useLocation();
  const { availableTours, startTour } = useTour();

  useEffect(() => {
    const unregister = registerCommands([
      {
        id: "navigate-pos",
        label: "Go to Point of Sale",
        description: "Jump to the primary sales workspace",
        group: "Navigation",
        shortcut: "G P",
        run: () => setLocation("/"),
      },
      {
        id: "navigate-admin",
        label: "Open Admin Dashboard",
        description: "Manage branches, users, and business settings",
        group: "Navigation",
        shortcut: "G A",
        run: () => setLocation("/admin"),
      },
      {
        id: "navigate-packages",
        label: "Review Packages",
        description: "View and configure subscription bundles",
        group: "Navigation",
        shortcut: "G K",
        run: () => setLocation("/packages"),
      },
      {
        id: "navigate-customers",
        label: "Customer Dashboard",
        description: "Check loyalty, rewards, and promotions",
        group: "Navigation",
        shortcut: "G C",
        run: () => setLocation("/customer-dashboard"),
      },
      {
        id: "navigate-driver",
        label: "Driver Console",
        description: "Monitor delivery routes and assignments",
        group: "Navigation",
        shortcut: "G D",
        run: () => setLocation("/driver"),
      },
    ]);

    return () => unregister();
  }, [registerCommands, setLocation]);

  useEffect(() => {
    if (!availableTours.length) return;
    const unregister = registerCommands(
      availableTours.map((tour) => ({
        id: `tour-${tour.id}`,
        label: `Start ${tour.title} tour`,
        description: "Launch guided onboarding",
        group: "Onboarding",
        shortcut: "?",
        run: () => startTour(tour.id),
      })),
    );
    return () => unregister();
  }, [availableTours, registerCommands, startTour]);

  const groupedCommands = useMemo<GroupedCommand[]>(() => {
    if (commands.length === 0) return [];
    const groups = new Map<string, GroupedCommand>();
    commands.forEach((command) => {
      const heading = command.group ?? "General";
      if (!groups.has(heading)) {
        groups.set(heading, { heading, items: [] });
      }
      groups.get(heading)?.items.push({
        id: command.id,
        label: command.label,
        description: command.description,
        shortcut: command.shortcut,
      });
    });
    return Array.from(groups.values()).map((group) => ({
      heading: group.heading,
      items: group.items.sort((a, b) => a.label.localeCompare(b.label)),
    }));
  }, [commands]);

  const leadingIcon = useMemo(() => {
    if (!isOpen) return null;
    const navigation = commands.find((cmd) => cmd.id === "navigate-pos");
    if (navigation) return <ShoppingCart className="size-5 text-primary" aria-hidden="true" />;
    return null;
  }, [commands, isOpen]);

  return (
    <CommandDialog open={isOpen} onOpenChange={setOpen}>
      <CommandInput placeholder="Search commands, pages, or tours…" />
      <CommandList>
        <CommandEmpty>No commands found. Try searching for a page or workflow.</CommandEmpty>
        {leadingIcon ? (
          <div className="flex items-center gap-2 px-4 py-2 text-[var(--text-sm)] text-muted-foreground">
            {leadingIcon}
            <span>Use ⌘K / Ctrl+K to toggle this palette from anywhere.</span>
          </div>
        ) : null}
        {groupedCommands.map((group, index) => (
          <div key={group.heading}>
            <CommandGroup heading={group.heading}>
              {group.items.map((item) => (
                <CommandItem key={item.id} onSelect={() => runCommand(item.id)} value={item.label}>
                  <GroupIcon heading={group.heading} />
                  <div className="flex flex-col">
                    <span>{item.label}</span>
                    {item.description ? (
                      <span className="text-[var(--text-xs)] text-muted-foreground">{item.description}</span>
                    ) : null}
                  </div>
                  {item.shortcut ? <CommandShortcut>{item.shortcut}</CommandShortcut> : null}
                </CommandItem>
              ))}
            </CommandGroup>
            {index < groupedCommands.length - 1 ? <CommandSeparator /> : null}
          </div>
        ))}
      </CommandList>
    </CommandDialog>
  );
}

function GroupIcon({ heading }: { heading: string }) {
  switch (heading) {
    case "Navigation":
      return <ShoppingCart className="mr-3 size-4 text-primary" aria-hidden="true" />;
    case "Quick Actions":
      return <Settings2 className="mr-3 size-4 text-secondary" aria-hidden="true" />;
    case "Onboarding":
      return <LifeBuoy className="mr-3 size-4 text-accent" aria-hidden="true" />;
    case "Customers":
      return <Users className="mr-3 size-4 text-primary" aria-hidden="true" />;
    case "Operations":
      return <ListChecks className="mr-3 size-4 text-secondary" aria-hidden="true" />;
    case "Packages":
      return <Package className="mr-3 size-4 text-accent" aria-hidden="true" />;
    default:
      return <ShoppingCart className="mr-3 size-4 text-muted-foreground" aria-hidden="true" />;
  }
}

export default CommandPalette;
