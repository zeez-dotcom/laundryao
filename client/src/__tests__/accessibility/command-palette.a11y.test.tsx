import { render } from "@testing-library/react";
import { axe } from "jest-axe";
import { useEffect } from "react";
import { describe, it, expect } from "vitest";
import CommandPalette from "@/components/navigation/CommandPalette";
import { CommandPaletteProvider, useCommandPalette } from "@/hooks/useCommandPalette";
import { TourProvider } from "@/components/onboarding/TourProvider";

class ResizeObserverMock {
  observe() {}
  unobserve() {}
  disconnect() {}
}

if (typeof window !== "undefined" && !window.ResizeObserver) {
  window.ResizeObserver = ResizeObserverMock as unknown as typeof window.ResizeObserver
}

if (!(globalThis as any).ResizeObserver) {
  ;(globalThis as any).ResizeObserver = ResizeObserverMock
}

function PaletteHarness() {
  const { setOpen, registerCommands } = useCommandPalette();

  useEffect(() => {
    const unregister = registerCommands([
      {
        id: "demo-command",
        label: "Demo command",
        description: "A sample action used for testing",
        group: "Testing",
        run: () => undefined,
      },
    ]);
    setOpen(true);
    return () => {
      unregister();
      setOpen(false);
    };
  }, [registerCommands, setOpen]);

  return <CommandPalette />;
}

describe("CommandPalette accessibility", () => {
  it("has no axe violations", async () => {
    const { container } = render(
      <CommandPaletteProvider>
        <TourProvider>
          <PaletteHarness />
        </TourProvider>
      </CommandPaletteProvider>,
    );

    expect(await axe(container)).toHaveNoViolations();
    expect(container).toMatchSnapshot();
  });
});
