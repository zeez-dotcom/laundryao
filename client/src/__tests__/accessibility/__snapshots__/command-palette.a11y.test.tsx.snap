// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommandPalette accessibility > has no axe violations 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
